pkg/kubelet/kubelet_pods.go

```diff
@@ -1431,6 +1432,18 @@ func getPhase(spec *v1.PodSpec, info []v1.ContainerStatus) v1.PodPhase {
 		// one container is running
 		return v1.PodRunning
 	case running == 0 && stopped > 0 && unknown == 0:
+		if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
```

Comment:
The one place this might break (reviewing the code) is if we are in syncPod and DeletionTimestamp is set - because this is called as part of generateAPIStatus and it happens before we start containers (so we could set a terminal phase here, but then start new containers, which we don't want).  Now that we have a tighter model in the Kubelet, that shouldn't happen anymore and in theory this is ok.

This makes me a bit nervous though, because there are other ways that pods can be killed that aren't deletion - primarily Eviction and graceful shutdown - and this makes the path through our code dependent on deletion, not on that higher level truth (roughly, that the pod is no longer desired to be running).  I *think* what I would suggest is that instead of relying on the deletion timestamp here, we rely on the kubelet's source of truth for whether a pod *will* be terminal (once a pod has been evicted, detected to be deleted, or reaches its natural conclusion).  That comes from the pod worker state machine (from the fact that you are in syncTerminatedPod) and can be passed down into this code.

So I was going to suggest that instead of you using this condition test here, we pass a boolean to both `generateAPIPodStatus` and `getPhase` which is `podIsTerminal bool` and pass it as true in syncTerminatedPod and false in the other methods.  That way we aren't tied to the deletion state here (deletion is just one trigger for the transition to terminal), and you inherit from the state machine of the pod worker (keeps the source of truth concentrated), and then that aligns TerminatePod (which doesn't need the check for deletion) to not need the check either (see my comment there).

I would probably say for cleanliness you *don't* have to keep passing Pod in (can revert to PodSpec only) in that case which reduces the change set a bit.

Reply:
You may also want to rename the PR and commit at that point to "Give terminal phase correctly to all pods that will not be restarted" (you're solving a real bug behind a phase gate, and you will improve the kubelet for everyone once we turn the gate on by default).

Given the closeness to 1.27 I'm ok with leaving the gate on, but in 1.28 I'd advocate for dropping the gate right away.

Reply:
I see, makes sense, I'm going to try. 

One thing, IIUC, this means that in some situations a pod that used to terminate in Failed, for example due to exceeding its `activeDeadlineSeconds` after the change could terminate in Succeeded, if all its containers are 0 (because they completed successfully before SIGKILL).

Reply:
If this is the case it can actually fix for us this issue: https://github.com/kubernetes/kubernetes/issues/115688

Reply:
Ah, done, it almost works, but again the test for MirrorPods fails, this time it gets the pod in Succeeded state. I guess I can exclude mirror pods from the logic in `getPhase`. WDYT?

Reply:
> One thing, IIUC, this means that in some situations a pod that used to terminate in Failed, for example due to exceeding its activeDeadlineSeconds after the change could terminate in Succeeded, if all its containers are 0 (because they completed successfully before SIGKILL).

I think that's desirable - since it's a race condition anyway we can safely get more precise.

Will look at the mirror pods logic real quick.

Reply:
Yeah #115668 is correct in saying that "the kubelet is *allowed* to set phase to failed for pods where all containers are 0", but in general "the kubelet *should* reduce the scenarios where that happens".  activeDeadlineSeconds actually describes its behavior as "best effort" to fail the pod:

> Optional duration in seconds the pod may be active on the node relative to StartTime before the system will actively try to mark it failed and kill associated containers

(we don't provide a hard guarantee that we will make it failed), so I think that's ok for here.  But container status is always subordinate to pod container status - clients should check phase first, containers second.  We should probably also have more validation in kubelet to report scenarios where that happens in our tests and then go try to fix the bugs (if pod = failed and container status == 0, report something in the test env).

Reply:
(wrong PR was referred)

Reply:
> (you're solving a real bug behind a phase gate, and you will improve the kubelet for everyone once we turn the gate on by default).

So, PodDisruptionConditions is currently in Beta, so enabled by default. However, I used it mostly cause it was motivated by the Jobs KEP. Since this is a real bug for Kubelet (I have opened this: https://github.com/kubernetes/kubernetes/issues/116410) I could happily drop the gate.

> Given the closeness to 1.27 I'm ok with leaving the gate on, but in 1.28 I'd advocate for dropping the gate right away.

I think dropping the gate now may actually speed up my work on this. WDYT?

Reply:
> One thing, IIUC, this means that in some situations a pod that used to terminate in Failed, for example due to exceeding its `activeDeadlineSeconds` after the change could terminate in Succeeded, if all its containers are 0 (because they completed successfully before SIGKILL).

Tested it, this is almost right, but it still ends in Failed, because the Succeeded phase is overwritten again here unconditionally: https://github.com/kubernetes/kubernetes/blob/bdf5af93e8a009f2c7eb66d0fa6d2e3ef5fdd061/pkg/kubelet/kubelet_pods.go#L1618. 

I think this is actually a mistake that we invoke the handlers even if the phase is terminal already.

Reply:
Ok, I tried that but we have this test "should allow activeDeadlineSeconds to be updated" (https://github.com/kubernetes/kubernetes/blob/90c3232de77aa0dd09b948ffdd27c575688fba8a/test/e2e/common/node/pods.go#L399) which starts to fail then, because the pause container exits with 0, and the pod is Succeeded, whereas the current assert expects Failed. I guess it is for another dedicated PR since this is a conformance test.

Reply:
Reviewing that test now to make sure I understand whether this is acceptable.

Reply:
I'm ok with dropping the gate, this is the correct improvement code.